returnn/torch/engine.py

```diff
@@ -359,9 +424,70 @@ def _run_step(self, extern_data_raw: Dict[str, torch.Tensor], *, train_flag: boo
+                self._pt_model.logger.set_runtime_stats_and_log()
+                self._pt_model.num_iterations += 1
+                self._pt_model.reducer.prepare_for_forward()
+
+            with torch.autograd.profiler.record_function("DistributedDataParallel.forward"):
+                if torch.is_grad_enabled() and self._pt_model.require_backward_grad_sync:
+                    assert self._pt_model.logger is not None
+                    self._pt_model.logger.set_runtime_stats_and_log()
+                    self._pt_model.num_iterations += 1
+                    self._pt_model.reducer.prepare_for_forward()
+
+                work = Join.notify_join_context(self._pt_model)
+                if work:
+                    self._pt_model.reducer._set_forward_pass_work_handle(
+                        work, self._pt_model._divide_by_initial_world_size  # type: ignore[arg-type]
+                    )
+
+                if torch.is_grad_enabled() and self._pt_model.reducer._rebuild_buckets():
+                    self._pt_model._has_rebuilt_buckets = True
+
+                if self._pt_model._check_sync_bufs_pre_fwd():
+                    self._pt_model._sync_buffers()
+
+                if self._pt_model._join_config.enable:
+                    # Notify joined ranks whether they should sync in backwards pass or not.
+                    self._pt_model._check_global_requires_backward_grad_sync(is_joined_rank=False)
+                with self._pt_model._inside_ddp_forward():
+                    with autocast(
+                        device_type=self._device, dtype=self._autocast_dtype
+                    ) if self._use_autocast else nullcontext():
+                        # If the flag use_DDP is enabled, we need to return the forward output in the train step function
+                        output = self._train_step_func(model=self._orig_model, extern_data=extern_data, **sentinel_kw)
+
+                if self._pt_model._check_sync_bufs_post_fwd():
+                    self._pt_model._sync_buffers()
+
+                if torch.is_grad_enabled() and self._pt_model.require_backward_grad_sync:
+                    self._pt_model.require_forward_param_sync = True
+                    # We'll return the output object verbatim since it is a freeform
+                    # object. We need to find any tensors in this object, though,
+                    # because we need to figure out which parameters were used during
+                    # this forward pass, to ensure we short circuit reduction for any
+                    # unused parameters. Only if `find_unused_parameters` is set.
+                    if self._pt_model.find_unused_parameters and not self._pt_model.static_graph:
+                        # Do not need to populate this for static graph.
+                        self._pt_model.reducer.prepare_for_backward(
+                            list(torch.nn.parallel.distributed._find_tensors(output))
```

Comment:
Use the loss instead. Access it via `rf.get_run_ctx()...`.

Reply:
I don't really understand what you mean. What do you mean by "the variable would be marked twice"? What does it mean to mark a variable?

To me it sounds like there is still some issue in the way we wrap the `forward` call.